PIPS/pip-1.md

```diff
@@ -0,0 +1,46 @@
+---
+pip: 1
+title: Permissionless LockProxies
+status: Draft
+author: John Wong (@jwheur)
+created: 2020-06-23
+---
+
+## Simple Summary
+The current LockProxy implementation requires an operator to control `setManagerProxy`, `bindProxyHash` and `bindAssetHash`.
+
+This PIP proposes an alternative implementation which does not require an operator.
+
+## Specification
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed.
+
+The `lock` function can be modified to require the following parameters:
```

Comment:
i have worry about user may fill in wrong info

Reply:
Hmm yea, I think we could have a function to register an asset:
```
function registerAsset(address fromAssetHash, uint64 toChainId, bytes memory targetProxyHash, bytes memory toAssetHash) {
    key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
    registry[key] = true;
}
```

Then the `lock` function can check this registry and raise an error if it does not exist.
So if the user types in the wrong info the transaction will fail.

Reply:
good idea! but whether it need a controller to register the asset?

Reply:
I think it should be okay without a controller, because any combination can be allowed

Reply:
if user bind a wrong mapping, it also can work right? whether this have some attack?

Reply:
Seems ok to me 🤔 

For example, if the token the user wants to transfer is DAI from Ethereum to Neo, any user can set up a Neo LockProxy and call `registerAsset(daiAssetHash, neoChainId, neoLockProxyAddr, neoDaiAssetHash)` on Ethereum.

If some attacker calls `registerAsset(daiAssetHash, neoChainId, neoLockProxyAddr, fakeNeoDaiAssetHash)`, it is still okay because when the user calls `lock` the user will send `lock(daiAssetHash, neoChainId, neoLockProxyAddr, neoDaiAssetHash, userNeoAddress, amount)`. 

There is only a risk if the attacker manages to get the user to also send `lock(daiAssetHash, neoChainId, neoLockProxyAddr, fakeNeoDaiAssetHash, userNeoAddress, amount)`, but if the attacker can do that then I think the user's funds are already compromised.

The `registry` would be helpful in case the user accidentally types `lock(daiAssetHash, neoChainId, neoLockProxyAddr, wrongNeoDaiAssetHash, userNeoAddress, amount)`, then `lock` will throw an error and help save the user's funds.

Reply:
if this method is called by asset contract when asset contract call the initialize method, i think is ok. it just can call once, and it will bind right mapping.

Reply:
actually I think we can remove `registerAsset` completely since `delegateAsset` will already update a `delegatedAssets` mapping that performs the same check, I'll remove it